NorthstarDLL/logging.cpp

```diff
@@ -12,257 +11,194 @@
 	}
 	else
 	{
-		try
-		{
-			// todo: might be good to delete logs that are too old
-			time_t time = std::time(nullptr);
-			tm currentTime = *std::localtime(&time);
-			std::stringstream stream;
+		freopen("CONOUT$", "w", stdout);
+		freopen("CONOUT$", "w", stderr);
+	}
 
-			stream << std::put_time(&currentTime, (GetNorthstarPrefix() + "/logs/nslog%Y-%m-%d %H-%M-%S.txt").c_str());
-			spdlog::default_logger()->sinks().push_back(std::make_shared<spdlog::sinks::basic_file_sink_mt>(stream.str(), false));
-			spdlog::flush_on(spdlog::level::info);
-		}
-		catch (...)
+	// this if statement is adapted from r5sdk
+	if (!strstr(GetCommandLineA(), "-noansiclr"))
+	{
+		DWORD dwMode = NULL;
+		HANDLE hOutput = GetStdHandle(STD_OUTPUT_HANDLE);
+
+		GetConsoleMode(hOutput, &dwMode);
+		dwMode |= ENABLE_PROCESSED_OUTPUT | ENABLE_VIRTUAL_TERMINAL_PROCESSING;
+
+		if (!SetConsoleMode(hOutput, dwMode)) // Some editions of Windows have 'VirtualTerminalLevel' disabled by default.
 		{
-			spdlog::error("Failed creating log file!");
-			MessageBoxA(
-				0, "Failed creating log file! Make sure the profile directory is writable.", "Northstar Warning", MB_ICONWARNING | MB_OK);
+			// If 'VirtualTerminalLevel' can't be set, just disable ANSI color, since it wouldnt work anyway.
+			spdlog::warn("could not set VirtualTerminalLevel. Disabling color output");
+			g_bSpdLog_UseAnsiColor = false;
 		}
 	}
 }
 
-void InitialiseLogging()
+void RegisterCustomSink(std::shared_ptr<CustomSink> sink)
 {
-	AllocConsole();
-
-	// Bind stdout to receive console output.
-	// these two lines are responsible for stuff to not show up in the console sometimes, from talking about it on discord
-	// apparently they were meant to make logging work when using -northstar, however from testing it seems that it doesnt
-	// work regardless of these two lines
-	// freopen("CONOUT$", "w", stdout);
-	// freopen("CONOUT$", "w", stderr);
-	spdlog::default_logger()->set_pattern("[%H:%M:%S] [%l] %v");
-}
+	for (auto& logger : loggers)
+	{
+		logger->custom_sinks_.push_back(sink);
+	}
+};
 
-ON_DLL_LOAD_RELIESON("engine.dll", EngineSpewFuncHooks, ConVar, (CModule module))
+void InitialiseLogging()
 {
-	AUTOHOOK_DISPATCH_MODULE(engine.dll)
+	// create a logger, and set it to default
+	NS::log::NORTHSTAR = std::make_shared<ColoredLogger>("NORTHSTAR", NS::Colors::NORTHSTAR, true);
+	NS::log::NORTHSTAR->sinks().clear();
+	loggers.push_back(NS::log::NORTHSTAR);
+	spdlog::set_default_logger(NS::log::NORTHSTAR);
+
+	// set whether we should use colour in the logs
+	g_bSpdLog_UseAnsiColor = !strstr(GetCommandLineA(), "-noansiclr");
+
+	// create our console sink
+	auto sink = std::make_shared<ExternalConsoleSink>();
+	// set the pattern
+	if (g_bSpdLog_UseAnsiColor)
+		// dont put the log level in the pattern if we are using colours, as the colour will show the log level
+		sink->set_pattern("[%H:%M:%S] [NAME] [LVL] %v");
```

Comment:
Idk if you have looked into this or not, but instead of doing string replacement on [NAME] and [LVL], it might be worth doing custom formatting flags for the sink or something?